src/main/java/org/opensearch/neuralsearch/stats/NeuralStatsInput.java

```diff
@@ -0,0 +1,168 @@
+     */
+    @Setter
+    private boolean flatten;
+
+    /**
+     * Builder constructor for creating NeuralStatsInput with specific filtering parameters.
+     *
+     * @param nodeIds node IDs to retrieve stats from
+     * @param eventStatNames event stats to retrieve
+     * @param stateStatNames state stats to retrieve
+     * @param includeMetadata whether to include metadata
+     * @param flatten whether to flatten keys
+     */
+    @Builder
+    public NeuralStatsInput(
+        Set<String> nodeIds,
+        EnumSet<EventStatName> eventStatNames,
+        EnumSet<StateStatName> stateStatNames,
+        boolean includeMetadata,
+        boolean flatten
+    ) {
+        this.nodeIds = nodeIds;
+        this.eventStatNames = eventStatNames;
+        this.stateStatNames = stateStatNames;
+        this.includeMetadata = includeMetadata;
+        this.flatten = flatten;
+    }
+
+    /**
+     * Default constructor that initializes with empty filters and default settings.
+     * By default, metadata is excluded and keys are not flattened.
+     */
+    public NeuralStatsInput() {
+        this.nodeIds = new HashSet<>();
+        this.eventStatNames = EnumSet.noneOf(EventStatName.class);
+        this.stateStatNames = EnumSet.noneOf(StateStatName.class);
+        this.includeMetadata = false;
+        this.flatten = false;
+    }
+
+    /**
+     * Constructor for stream input
+     *
+     * @param input the StreamInput to read data from
+     * @throws IOException if there's an error reading from the stream
+     */
+    public NeuralStatsInput(StreamInput input) throws IOException {
+        nodeIds = input.readBoolean() ? new HashSet<>(input.readStringList()) : new HashSet<>();
+        eventStatNames = input.readBoolean() ? input.readEnumSet(EventStatName.class) : EnumSet.noneOf(EventStatName.class);
+        stateStatNames = input.readBoolean() ? input.readEnumSet(StateStatName.class) : EnumSet.noneOf(StateStatName.class);
+        includeMetadata = input.readBoolean();
+        flatten = input.readBoolean();
+    }
+
+    /**
+     * Serializes this object to a StreamOutput.
+     *
+     * @param out the StreamOutput to write data to
+     * @throws IOException If there's an error writing to the stream
+     */
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeOptionalStringCollection(nodeIds);
+        writeOptionalEnumSet(out, eventStatNames);
+        writeOptionalEnumSet(out, stateStatNames);
+        out.writeBoolean(includeMetadata);
+        out.writeBoolean(flatten);
+    }
+
+    /**
+     * Helper method to write an EnumSet to a StreamOutput with a presence flag.
+     *
+     * @param out the StreamOutput to write to
+     * @param set the EnumSet to write
+     * @throws IOException If there's an error writing to the stream
+     */
+    private void writeOptionalEnumSet(StreamOutput out, EnumSet<?> set) throws IOException {
```

Comment:
Might be good to have corresponding `readOptionalEnumSet()`
Better if we can add this method inside core :)

Reply:
Until core is ready, we can have the method in this class temporary.